session/sql_store.go

```diff
@@ -0,0 +1,839 @@
+	// with any of our existing IDs.
+	for {
+		// Generate a new private key and Alias pair.
+		privKey, alias, err := NewSessionPrivKeyAndID()
+		if err != nil {
+			return ID{}, nil, err
+		}
+
+		// Check that no such legacy Alias exits.
+		_, err = db.GetSessionByAlias(ctx, alias[:])
+		if errors.Is(err, sql.ErrNoRows) {
+			return alias, privKey, nil
+		} else if err != nil {
+			return ID{}, nil, fmt.Errorf("unable to get "+
+				"session: %w", err)
+		}
+
+		continue
+	}
+}
+
+// GetSession returns the session with the given legacy Alias.
+//
+// NOTE: This is part of the Store interface.
+func (s *SQLStore) GetSession(ctx context.Context, alias ID) (*Session, error) {
+	var (
+		readTxOpts = db.NewQueryReadTx()
+		sess       *Session
+	)
+	err := s.db.ExecTx(ctx, &readTxOpts, func(db SQLQueries) error {
+		dbSess, err := db.GetSessionByAlias(ctx, alias[:])
+		if errors.Is(err, sql.ErrNoRows) {
+			return ErrSessionNotFound
+		} else if err != nil {
+			return fmt.Errorf("unable to get session: %w", err)
+		}
+
+		sess, err = unmarshalSession(ctx, s.db, dbSess)
+		if err != nil {
+			return fmt.Errorf("unable to unmarshal session: %w",
+				err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return sess, nil
+}
+
+// GetGroupID will return the legacy group Alias for the given legacy session Alias.
+//
+// NOTE: This is part of the AliasToGroupIndex interface.
+func (s *SQLStore) GetGroupID(ctx context.Context, sessionID ID) (ID, error) {
+	var (
+		readTxOpts    = db.NewQueryReadTx()
+		legacyGroupID ID
+	)
+	err := s.db.ExecTx(ctx, &readTxOpts, func(db SQLQueries) error {
+		// Get the session using the legacy Alias.
+		sess, err := db.GetSessionByAlias(ctx, sessionID[:])
+		if err != nil {
+			return ErrUnknownGroup
+		}
+
+		if !sess.GroupID.Valid {
+			return fmt.Errorf("session does not have a group Alias")
+		}
+
+		// Get the legacy group Alias using the session group Alias.
+		legacyGroupIDB, err := db.GetAliasBySessionID(
+			ctx, sess.GroupID.Int64,
+		)
+		if err != nil {
+			return err
+		}
+
+		legacyGroupID, err = IDFromBytes(legacyGroupIDB)
+
+		return err
+	})
+	if err != nil {
+		return ID{}, err
+	}
+
+	return legacyGroupID, nil
+}
+
+// GetSessionIDs will return the set of legacy session IDs that are in the
+// group with the given legacy Alias.
+//
+// NOTE: This is part of the AliasToGroupIndex interface.
+func (s *SQLStore) GetSessionIDs(ctx context.Context, legacyGroupID ID) ([]ID,
+	error) {
+
+	var (
+		readTxOpts = db.NewQueryReadTx()
+		sessionIDs []ID
+	)
+	err := s.db.ExecTx(ctx, &readTxOpts, func(db SQLQueries) error {
+		groupID, err := db.GetSessionIDByAlias(ctx, legacyGroupID[:])
+		if err != nil {
+			return fmt.Errorf("unable to get session Alias: %v",
+				err)
+		}
+
+		sessIDs, err := db.GetSessionAliasesInGroup(
+			ctx, sql.NullInt64{
+				Int64: groupID,
+				Valid: true,
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("unable to get session IDs: %v", err)
+		}
```

Comment:
Just noting that original logic actually seems to error if no sessions for the legacyGroupID is found, while this would return an emprty list. But given that I think the new logic is actually more correct, and I don't see a scenario where we'd hit that anyways, I think we should probably keep as is.